File: 2015/03_Perfectly_Spherical_Houses_in_a_Vacuum/part_1/main.js
/**
 * Author: happ_2h https://github.com/happ-2h
 *
 * Solution for Advent of Code 2015 Day 3.1
 */

let input = null;

try {
  input = require('fs').readFileSync("../input.txt", "utf-8").split("");
} catch (err) {
  console.error(err);
}

if (input !== null) {
  const santa = { x: 0, y: 0};        // Coordinates
  const visited = new Set();          // Houses visited (unique)
  visited.add(JSON.stringify(santa)); // Save starting point

  while(input.length > 0) {
    const dir = input.shift();

    switch(dir) {
      case "^": --santa.y; break;
      case "v": ++santa.y; break;
      case "<": --santa.x; break;
      case ">": ++santa.x; break;
    }

    visited.add(JSON.stringify(santa));
  }

  console.log(visited.size);
}
else console.error("Failed to read input.txt");

File: 2015/03_Perfectly_Spherical_Houses_in_a_Vacuum/part_1/main.kt
/**
 * Author: happ_2h https://github.com/happ-2h
 *
 * Solution for Advent of Code 2015 Day 3.1
 */

import java.io.File

// Keep track of Cartesian coordinates
data class Santa(var x: Int, var y: Int)

fun main(args: Array<String>) {
  val path = File("../input.txt").readText()

  // Create and save starting point
  val santa: Santa = Santa(0, 0)
  val visited: MutableSet<String> = mutableSetOf(santa.toString())

  // Move based on input
  path.forEach {
    when(it) {
      '^' -> --santa.y
      'v' -> ++santa.y
      '<' -> --santa.x
      '>' -> ++santa.x
      else -> println("Unknown $it")
    }

    // Save result (unique)
    visited.add(santa.toString())
  }

  println(visited.size)
}